desc/integrals/basis.py

```diff
@@ -58,6 +85,627 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+        Chebyshev spectral resolution.
+    domain : (float, float)
+        Domain for y coordinates.
+    lobatto : bool
+        Whether ``f`` was sampled on the Gauss-Lobatto (extrema-plus-endpoint)
+        instead of the interior roots grid for Chebyshev points.
+
+    """
+
+    def __init__(self, f, domain=(-1, 1), lobatto=False):
+        """Interpolate Fourier-Chebyshev series to ``f``.
+
+        Parameters
+        ----------
+        f : jnp.ndarray
+            Shape (..., M, N).
+            Samples of real function on the ``FourierChebyshevSeries.nodes`` grid.
+        domain : (float, float)
+            Domain for y coordinates. Default is [-1, 1].
+        lobatto : bool
+            Whether ``f`` was sampled on the Gauss-Lobatto (extrema-plus-endpoint)
+            instead of the interior roots grid for Chebyshev points.
+
+        """
+        self.M = f.shape[-2]
+        self.N = f.shape[-1]
+        errorif(domain[0] > domain[-1], msg="Got inverted domain.")
+        self.domain = tuple(domain)
+        errorif(lobatto, NotImplementedError, "JAX hasn't implemented type 1 DCT.")
+        self.lobatto = bool(lobatto)
+        self._c = FourierChebyshevSeries._transform(f, self.lobatto)
+
+    @staticmethod
+    def _transform(f, lobatto):
+        N = f.shape[-1]
+        return rfft(
+            dct(f, type=2 - lobatto, axis=-1) / (N - lobatto),
+            axis=-2,
+            norm="forward",
+        )
+
+    @staticmethod
+    def nodes(M, N, L=None, domain=(-1, 1), lobatto=False):
+        """Tensor product grid of optimal collocation nodes for this basis.
+
+        Parameters
+        ----------
+        M : int
+            Grid resolution in x direction. Preferably power of 2.
+        N : int
+            Grid resolution in y direction. Preferably power of 2.
+        L : int or jnp.ndarray
+            Optional, resolution in radial direction of domain [0, 1].
+            May also be an array of coordinates values. If given, then the
+            returned ``coords`` is a 3D tensor-product with shape (L * M * N, 3).
+        domain : tuple[float]
+            Domain for y coordinates. Default is [-1, 1].
+        lobatto : bool
+            Whether to use the Gauss-Lobatto (Extrema-plus-Endpoint)
+            instead of the interior roots grid for Chebyshev points.
+
+        Returns
+        -------
+        coords : jnp.ndarray
+            Shape (M * N, 2).
+            Grid of (x, y) points for optimal interpolation.
+
+        """
+        x = fourier_pts(M)
+        y = cheb_pts(N, domain, lobatto)
+        if L is None:
+            coords = (x, y)
+        else:
+            if isposint(L):
+                L = jnp.flipud(jnp.linspace(1, 0, L, endpoint=False))
+            coords = (jnp.atleast_1d(L), x, y)
+        coords = tuple(map(jnp.ravel, jnp.meshgrid(*coords, indexing="ij")))
+        return jnp.column_stack(coords)
+
+    def evaluate(self, M, N):
+        """Evaluate Fourier-Chebyshev series.
```

Comment:
maybe specify that this is evaluating on tensor product grid, not arbitary points (to distinguish from `Basis.evaluate`).

Also, does this need `L` as an option?

Reply:
Ok; no for L because we can't interpolate to new rho values. It's a 2D basis